torch_xla/csrc/runtime/stablehlo_composite_helper.cc

```diff
@@ -0,0 +1,404 @@
+          // Terminal condition: constant
+          scope_ops_setvec.insert(def_op);
+        } else {
+          processing.push_back(def_op);
+        }
+      }
+    }
+    // Sorts all ops within the boundary by their line numbers in the input
+    // MLIR. The ops will be duplicated to the impl function following this
+    // order.
+    auto scope_ops = scope_ops_setvec.takeVector();
+    for (auto& op : scope_ops) {
+      if (!op_line_num.contains(op)) {
+        return;
+      }
+    }
+    std::sort(scope_ops.begin(), scope_ops.end(),
+              [&op_line_num](const auto& a, const auto& b) {
+                return op_line_num.at(a) < op_line_num.at(b);
+              });
+
+    // Sorts boundary args by their positions. Note that the args of the
+    // composite and impl function may be more than the boundary inputs, because
+    // the MLIR is lowered from the functionalized graph and additional args may
+    // be Pytorch constants. In such case the position of those args would be
+    // undetermined, while they would always come after boundary inputs.
+    auto arg_pos_pairs = arg_pos_setvec.takeVector();
+    std::stable_sort(
+        arg_pos_pairs.begin(), arg_pos_pairs.end(),
+        [](const auto& a, const auto& b) { return a.second < b.second; });
+    llvm::SmallVector<mlir::Value> args;
+    args.reserve(arg_pos_pairs.size());
+    for (auto& [arg, unused] : arg_pos_pairs) {
+      args.push_back(arg);
+    }
+
+    // Creates composite impl function and duplicates all ops within the
+    // boundary in the function.
+    llvm::SmallVector<mlir::Location> arg_locs;
+    llvm::SmallVector<mlir::Type> arg_types,
+        result_types(op->getResultTypes().begin(), op->getResultTypes().end());
+    for (auto& arg : args) {
+      arg_types.push_back(arg.getType());
+      arg_locs.push_back(arg.getLoc());
+    }
+
+    mlir::func::FuncOp impl_func = builder.create<mlir::func::FuncOp>(
+        module_op.getLoc(),
+        absl::StrCat(output_metadata.boundary_id(), ".impl"),
+        mlir::FunctionType::get(context, arg_types, result_types));
+    mlir::IRMapping mapping;
+    builder.createBlock(&impl_func.getBody(), impl_func.begin(), arg_types,
+                        arg_locs);
+    for (const auto& arg : llvm::enumerate(args)) {
+      mapping.map(arg.value(), impl_func.getArgument(arg.index()));
+    }
+    for (mlir::Operation* original_op : scope_ops) {
+      mlir::Operation* cloned_op = builder.clone(*original_op, mapping);
+      mapping.map(original_op, cloned_op);
+    }
+    builder.create<mlir::func::ReturnOp>(impl_func.getBody().getLoc(),
+                                         mapping.lookup(op)->getResults());
+
+    // Adds the new function to symbol table.
+    mlir::SymbolTable symbol_table(module_op);
+    impl_func.setPrivate();
+    symbol_table.insert(impl_func);
+
+    builder.setInsertionPointAfter(op);
+    llvm::SmallVector<mlir::NamedAttribute> call_attrs{
```

Comment:
Would suggest making this a separate method: `buildCompositeAsCustomCall`. Will make future refactoring to composite much easier.